src/main/java/ch/njol/skript/entity/MinecartData.java

```diff
@@ -12,142 +11,115 @@
 	}
-	
-	@SuppressWarnings("null")
+
 	@Override
-	protected boolean init(final Literal<?>[] exprs, final int matchedPattern, final ParseResult parseResult) {
-		type = MinecartType.values()[matchedPattern];
+	protected boolean init(Literal<?>[] exprs, int matchedPattern, ParseResult parseResult) {
+		if (matchedPattern == 0)
+			return true;
+
+		type = MinecartType.values()[--matchedPattern];
+		isSupertype = false;
 		return true;
 	}
-	
-	@SuppressWarnings("null")
+
 	@Override
-	protected boolean init(final @Nullable Class<? extends Minecart> c, final @Nullable Minecart e) {
-		final MinecartType[] ts = MinecartType.values();
-		for (int i = ts.length - 1; i >= 0; i--) {
-			final Class<?> mc = ts[i].c;
-			if (mc == null)
-				continue;
-			if (e == null ? mc.isAssignableFrom(c) : mc.isInstance(e)) {
-				type = ts[i];
-				return true;
+	protected boolean init(@Nullable Class<? extends Minecart> entityClass, @Nullable Minecart entity) {
+		for (MinecartType t : MinecartType.values()) {
+			Class<?> mc = t.entityClass;
+			if (entity == null ? mc.isAssignableFrom(entityClass) : mc.isInstance(entity)) {
+				isSupertype = false;
+				type = t;
+				break;
 			}
 		}
-		assert false;
-		return false;
+		return true;
 	}
-	
+
 	@Override
-	public void set(final Minecart entity) {}
-	
+	public void set(Minecart entity) {}
+
 	@Override
-	public boolean match(final Minecart entity) {
-		if (type == MinecartType.NORMAL && type.c == Minecart.class) // pre-1.5
-			return !(entity.getClass().equals(Utils.classForName("org.bukkit.entity.StorageMinecart"))
-					|| entity.getClass().equals(Utils.classForName("org.bukkit.entity.PoweredMinecart")));
-		return type.c != null && type.c.isInstance(entity);
+	public boolean match(Minecart entity) {
+		return isSupertype || type.entityClass.isInstance(entity);
 	}
-	
+
 	@Override
 	public Class<? extends Minecart> getType() {
-		return type.c != null ? type.c : Minecart.class;
+		return isSupertype ? Minecart.class : type.entityClass;
 	}
-	
+
 	@Override
 	protected int hashCode_i() {
 		return type.hashCode();
 	}
-	
-	@Override
-	protected boolean equals_i(final EntityData<?> obj) {
-		if (!(obj instanceof MinecartData))
-			return false;
-		final MinecartData other = (MinecartData) obj;
-		return type == other.type;
-	}
-	
-//		return type.name();
+
 	@Override
-	protected boolean deserialize(final String s) {
-		try {
-			type = MinecartType.valueOf(s);
-			return true;
-		} catch (final IllegalArgumentException e) {
-			return false;
-		}
+	protected boolean equals_i(EntityData<?> obj) {
+		return obj instanceof MinecartData minecartData && (isSupertype || type == minecartData.type);
 	}
-	
+
 	@Override
-	public boolean isSupertypeOf(final EntityData<?> e) {
-		if (e instanceof MinecartData)
-			return type == MinecartType.ANY || ((MinecartData) e).type == type;
-		return false;
+	public boolean isSupertypeOf(EntityData<?> other) {
+		return other instanceof MinecartData && isSupertype;
```

Comment:
```suggestion
		return other instanceof MinecartData minecartData && (isSupertype || type == other.type);
```
`#isSupertypeOf` should check for supertype  or same type